File: src/SafetyChecker.cpp
#include "SafetyChecker.h"
#include <unordered_map>
#include <utility>

bool SafetyChecker::validate(const SignalPhase &phase) const {
    int greenCount = 0;
    for (const auto &entry : phase.getStates()) {
        if (entry.second == LightState::Green) {
            ++greenCount;
            if (greenCount > 1) {
                return false; // Disallow multiple greens in this simple safety check.
            }
        }
    }
    return true;
}

SignalPhase SafetyChecker::fallbackPhase(const std::vector<std::string> &approaches) const {
    std::unordered_map<std::string, LightState> states;
    for (const auto &approachId : approaches) {
        states[approachId] = LightState::FlashingAmber;
    }
    return SignalPhase("fallback-flashing-amber", 3, std::move(states));
}

File: python/src/safety_checker.py
from typing import List, Dict
from .signal_phase import SignalPhase
from .light_state import LightState

class SafetyChecker:
    def validate(self, phase: SignalPhase) -> bool:
        # Simple safety: accept phases as-is
        return True

    def fallbackPhase(self, approach_ids: List[str]) -> SignalPhase:
        states: Dict[str, LightState] = {a: LightState.AllRed for a in approach_ids}
        return SignalPhase("fallback-all-red", 1, states)
